Kitodo/src/main/java/org/kitodo/production/helper/tasks/HierarchyMigrationTask.java

```diff
@@ -0,0 +1,358 @@
+                .filter(metadata -> metadata.getKey().equals("CurrentNo")).filter(MetadataEntry.class::isInstance)
+                .map(MetadataEntry.class::cast).map(MetadataEntry::getValue).map(Integer::valueOf).findFirst()
+                .orElse(null);
+        return currentNo;
+    }
+
+    /**
+     * Links a child process in an existing parent process.
+     *
+     * @param childProcess
+     *            child process to link
+     * @param parentData
+     *            a data object with the ID of the parent process and the
+     *            current numbers of the child processes already linked with the
+     *            parent
+     */
+    private static void linkProcessInParent(Process childProcess, List<Integer> parentData)
+            throws IOException, DAOException, DataException {
+
+        URI metadataFilePath = fileService.getMetadataFilePath(childProcess);
+        Integer currentNo = convertChildMetsFile(metadataFilePath);
+        Process parentProcess = processService.getById(parentData.get(0));
+        int insertionPosition = calculateInsertionPosition(parentData, currentNo);
+        MetadataEditor.addLink(parentProcess, Integer.toString(insertionPosition), childProcess.getId());
+        parentData.add(insertionPosition + 1, currentNo);
+        linkInDatabase(parentProcess, childProcess);
+    }
+
+    /**
+     * Calculates the point at which the child process must be inserted in the
+     * parent hierarchy.
+     *
+     * @param parentData
+     *            a data object with the ID of the parent process (here unused)
+     *            and the current numbers of the child processes already linked
+     *            with the parent, which may be {@code null}
+     * @param currentNo
+     *            the current number of the child process to link, may be
+     *            {@code null}
+     * @return the insertion position
+     */
+    private static int calculateInsertionPosition(List<Integer> parentData, Integer currentNo) {
+        int currentNumber = Objects.isNull(currentNo) ? Integer.MIN_VALUE : currentNo;
+        int insertionPosition = 0;
+        for (int index = 1; index < parentData.size(); index++) {
+            int comparee = Objects.isNull(parentData.get(index)) ? Integer.MIN_VALUE : parentData.get(index);
+            if (currentNumber >= comparee) {
+                insertionPosition++;
+            } else {
+                break;
+            }
+        }
+        return insertionPosition;
+    }
+
+    /**
+     * Renames the anchor file. It is clear that this has been migrated.
+     *
+     * @param process
+     *            process to migrate
+     */
+    private static void renameAnchorFile(Process process) throws IOException {
+        URI anchorFile = fileService.createAnchorFile(fileService.getMetadataFilePath(process));
+        fileService.renameFile(anchorFile, "meta_anchor.migrated");
```

Comment:
Is kitodo still working with this renamed anchor file?
@henning-gerhardt  what do you think?

Reply:
I don't understand why this file is still existing after migration and why it get a new name. Maybe @matthias-ronge should explain this in more detail. Maybe direct in German between @Kathrin-Huber and me.

Reply:
Simple, I‘m not deleting the file for safety‘s sake if anything goes wrong, but I‘m renaming it so that the migration cannot be accidentally repeated (principle of [idempotence](https://de.wikipedia.org/wiki/Idempotenz#Informatik)). I had followed the same procedure for the newspaper process migration. In principle, the file could also be deleted, but then there is no going back if something went wrong during the migration. So, consider this file as a backup, similar to backup file rotation.